include/Dialect/Polynomial/IR/NumberTheory.h

```diff
@@ -0,0 +1,15 @@
+#include "llvm/include/llvm/ADT/APInt.h"  // from @llvm-project
+
+namespace mlir::heir::polynomial {
+
+/// Returns true if `root` is a primitive nth root modulo `cmod`.
+bool isPrimitiveNthRootOfUnity(const ::llvm::APInt &root, unsigned n,
+                               const ::llvm::APInt &cmod);
+
+/// Returns true if `root` is a primitive 2nth root modulo `cmod`.
+inline bool isPrimitive2nthRootOfUnity(const ::llvm::APInt &root, unsigned n,
```

Comment:
@inbelic TBH, this method might not even be necessary, just call it Nth root and rely on the user to pass in 2*n when needed.

Reply:
It should be straightforward enough to do so. I can modify that when I cherry-pick it